File: teoria/grafos/busca/bfs_matriz.cpp
#include <bits/stdc++.h> // Inclui todas as bibliotecas padrão do C++. 

#define MAX 505 // Define um valor constante MAX como 505.
using namespace std; // Usa o namespace padrão std.

int M[MAX][MAX]; // Declara uma matriz de adjacência M para o grafo.
int visitados[MAX]; // Declara um array para marcar os vértices visitados.
int V; // Declara uma variável para o número de vértices.

// Função BFS para percorrer o grafo a partir de um vértice de origem.
void bfs(int origem)
{
    queue<int> fila; // Declara uma fila para a BFS.
    fila.push(origem); // Insere o vértice de origem na fila.
    visitados[origem] = 1; // Marca o vértice de origem como visitado.

    while (!fila.empty()) // Enquanto a fila não estiver vazia.
    {
        int v_atual = fila.front(); // Pega o vértice na frente da fila.
        fila.pop(); // Remove o vértice da fila.

        printf("Visitando vertice %d\n", v_atual); // Imprime o vértice sendo visitado.

        for (int vz = 0; vz <= V; vz++) // Itera por todos os vértices.
            if (visitados[vz] == 0 and M[v_atual][vz] == 1) // Se o vértice vz não foi visitado e é adjacente ao vértice atual.
            {
                fila.push(vz); // Adiciona o vértice vz à fila.
                visitados[vz] = 1; // Marca o vértice vz como visitado.
            }
    }
}

int main()
{
    int A; // Declara uma variável para o número de arestas.
    memset(M, 0, sizeof(M)); // Inicializa a matriz de adjacência M com zeros.
    memset(visitados, 0, sizeof(visitados)); // Inicializa o array de visitados com zeros.
    scanf("%d%d", &V, &A); // Lê o número de vértices e arestas.
    int v1, v2, peso; // Declara variáveis para ler as arestas.

    for (int k = 0; k < A; k++) // Loop para ler todas as arestas.
    {
        scanf("%d%d", &v1, &v2); // Lê uma aresta entre v1 e v2.
        M[v1 - 1][v2 - 1] = 1; // Marca a aresta na matriz de adjacência.
        M[v2 - 1][v1 - 1] = 1; // Marca a aresta na matriz de adjacência (grafo não direcionado).
    }

    bfs(0); // Chama a BFS a partir do vértice 0.

    for (int i = 0; i < V; i++) // Loop para imprimir os vértices visitados.
        printf("%d ", visitados[i]); // Imprime o array de visitados.

    return 0; // Retorna 0 para indicar que o programa terminou com sucesso.
}

File: teoria/grafos/busca/bfs_matriz.py
INF = int(1e9)  # INF = 1B, usado como valor sentinela para vértices não visitados

def bfs(origem, M, V):
    visitados = [INF] * (V+1)  # Inicializa todos os vértices como não visitados
    visitados[origem] = 1      # Marca o vértice de origem como visitado
    fila = [origem]            # Inicializa a fila de BFS com o vértice de origem

    while len(fila) != 0:      # Enquanto a fila não estiver vazia
        v_atual = fila[0]      # Obtém o primeiro vértice da fila
        del fila[0]            # Remove o primeiro vértice da fila

        print('Visitando o vertice %d' % v_atual)  # Imprime o vértice atual sendo visitado

        for vz in range(V+1):  # Itera sobre todos os vértices vizinhos
            if visitados[vz] == INF and M[v_atual][vz] == 1:  # Se o vizinho não foi visitado e há aresta
                visitados[vz] = 1  # Marca o vizinho como visitado
                fila.append(vz)    # Adiciona o vizinho à fila

    return visitados  # Retorna a lista de vértices visitados

# Leitura de entradas
L = input().split()
V, A = int(L[0]), int(L[1])  # Número de vértices e arestas
M = [[0 for i in range(V+1)] for j in range(V+1)]  # Inicializa a matriz de adjacência

for i in range(A):
    L = input().split()
    A, B = int(L[0])-1, int(L[1])-1  # Lê os vértices das arestas, ajusta para índices baseados em zero
    M[A][B] = 1  # Marca a aresta na matriz de adjacência
    M[B][A] = 1  # Grafo não direcionado, marca a aresta de volta também
    
print(*M, sep='\n')

print(bfs(0, M, V))  # Chama a função BFS a partir do vértice 0 e imprime o resultado
